### PR TITLE
Add cloud-controller-manager support for Kubernetes cluster

### PR DESCRIPTION
**Pull request by karataliu**
**What this PR does / why we need it**:
Kubernetes v1.8 supports [cloud-controller-manager ](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/) which runs cloud provider related logic, and it is planned in v1.10 that using cloud-controller-manager becomes the only way to interact with clouds.

This PR adds an option 'UseCloudControllerManager' in apimodel for Kubernetes v1.8 version. If turned on, the cluster will be using cloud controller manager mode.

Currently it uses the common image 'cloud-controller-manager-amd64:v1.8.0' which includes all cloud provider implementation. Will later switch to azure dedicated cloud-controller-manager image once ready.

One TODO item is RBAC support, which pends on upstream Kubernetes#53511 .

Please find more details here: https://github.com/kubernetes/kubernetes/issues/50752

**Which issue this PR fixes**

**Special notes for your reviewer**:

**Release note**:
```release-note
Add cloud-controller-manager support for Kubernetes cluster
```

### CODE REVIEW

**Comment by karataliu** [comment]
cc @anhowe @jdumars  This PR is for supporting standalone Azure cloud provider from acs-engine side. Could you please take a look? Thanks!

File: parts/kubernetesmaster-cloud-controller-manager.yaml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      command: 
+        - "cloud-controller-manager"
+        - "--kubeconfig=/var/lib/kubelet/kubeconfig"
+        - "--allocate-node-cidrs=<allocateNodeCidrs>"
+        - "--cluster-cidr=<kubeClusterCidr>"
+        - "--cluster-name=<masterFqdnPrefix>"
+        - "--cloud-provider=azure"
+        - "--cloud-config=/etc/kubernetes/azure.json"
+        - "--leader-elect=true"
+# TODO: RBAC support
```
Comment (colemickens) [inline parts/kubernetesmaster-cloud-controller-manager.yaml]: What's needed or outstanding for enabling RBAC?

File: parts/kubernetesmastercustomdata.yml [hunk 2]
```diff
@@ -302,6 +316,16 @@ write_files:
 {{end}}
     sed -i "s|<etcdApiVersion>|{{ .OrchestratorProfile.GetAPIServerEtcdAPIVersion }}|g" "/etc/kubernetes/manifests/kube-apiserver.yaml"
 
+{{if UseCloudControllerManager }}
+    sed -i "s|<kubernetesCcmImageSpec>|{{WrapAsVariable "kubernetesCcmImageSpec"}}|g; s|<masterFqdnPrefix>|{{WrapAsVariable "masterFqdnPrefix"}}|g; s|<allocateNodeCidrs>|{{WrapAsVariable "allocateNodeCidrs"}}|g; s|<kubeClusterCidr>|{{WrapAsVariable "kubeClusterCidr"}}|g; s|<kubernetesCtrlMgrRouteReconciliationPeriod>|{{WrapAsVariable "kubernetesCtrlMgrRouteReconciliationPeriod"}}|g" \
+      /etc/kubernetes/manifests/cloud-controller-manager.yaml
+
+    sed -i "/--\(cloud-config\|cloud-provider\|route-reconciliation-period\)=/d" \
```
Comment (colemickens) [inline parts/kubernetesmastercustomdata.yml]: Is this stripping off `cloud-provider` and `cloud-config` from apiserver? Doesn't that also need to be stripped off of `controller-manager` too?

File: parts/kubernetesagentcustomdata.yml [hunk 3]
```diff
@@ -125,6 +125,11 @@ write_files:
     KUBELET_FIX_43704_3=""
   {{end}}
 {{end}}
+{{if UseCloudControllerManager }}
+    CLOUD_PROVIDER=external
+{{else}}
+    CLOUD_PROVIDER=azure
+{{end}}
 
 - path: "/etc/systemd/system/kubelet.service"
```
Comment (colemickens) [inline parts/kubernetesagentcustomdata.yml]: Is there no change needed for `kubelet.service` to include the `--provider-id` flag?

**Comment by colemickens** [comment]
What's the status of this @karataliu? What level of functionality does the cluster have when this is enabled? I'm guessing volumes don't work, but does networking + load balancing?

I'm trying to understand how many of the items mentioned in the upstream issue need to happen before CCM + Azure is functional for most non-volume use-cases. Thanks.

File: parts/kubernetesmastercustomdata.yml (hunk 2 above)
Comment (karataliu) [reply parts/kubernetesmastercustomdata.yml]: Yeah, both need to be updated. LIne323~324 is for kube-controller-manager, Line325~326 is for apiserver.

File: parts/kubernetesmaster-cloud-controller-manager.yaml (hunk 1 above)
Comment (karataliu) [reply parts/kubernetesmaster-cloud-controller-manager.yaml]: This is actually meant for enabling 'use-service-account-credentials', which does not work now, pending on kubernetes#53511

File: parts/kubernetesagentcustomdata.yml (hunk 3 above)
Comment (karataliu) [reply parts/kubernetesagentcustomdata.yml]: provider-id can be detected now:
https://github.com/kubernetes/kubernetes/pull/50730
https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/azure/azure_instances.go#L103

**Comment by karataliu** [comment]
@colemickens Yes, volumes feature does not work. I've tried with e2e conformance tests and most cases would pass excepted for the volume ones.

The upstream issue is more about migration work, Azure provider running with cloud-controller-manager could already work on a cluster. But since the whole feature is under beta stage (kubernetes/features#88), it is not recommended for production use.

**Comment by colemickens** [comment]
Thanks for the info. Change seems good to me. I made a similar change in
Tectonic-Installer and everything seemed to work (at least,the sunset I
expected to work). (Granted the network setup in Tectonic-Installer is
Calico and not kubelet...)

**Comment by karataliu** [comment]
Could someone please take a look? Thanks.

**Comment by colemickens** [comment]
(FWIW, this looks good to me. And in a few weeks hopefully, there should be alpha support for CCM under hyperkube as well, so there won't need to be a separate image/process... at least until CCM totally splits out of the main repo and is yanked back out of hyperkube, but that could be a while).

**Comment by karataliu** [comment]
@colemickens Thanks, I saw the ongoing work for hyperkube incorporating CCM.
For the change here, a dedicate parameter for CCM image is still required as it could be set to an out of band CCM release. Also during CI test, a new CCM image is built and tested every time.

**Comment by jackfrancis** [comment]
Thanks for your patience @karataliu, and apologies for the rebase needs. Not a few things have changed in the conflicting surface area, I'd be happy to spawn a new PR with this work if that's easier.

**Comment by karataliu** [comment]
@jackfrancis Updated, please take a look. Thanks.

**Comment by jackfrancis** [comment]
Thanks @karataliu ! Running E2E now.

**Comment by karataliu** [comment]
In fact the default value should be dependent on k8s version, and as planned for 1.10+ release it should only have true value.

Anyway it is good to switch to ternary here, I've updated accordingly.

**Pull request merged**